From be4af54a04251280af796fd387653794e1d52915 Mon Sep 17 00:00:00 2001
From: Alex Naidis <alex.naidis@linux.com>
Date: Thu, 13 Oct 2016 19:02:15 +0200
Subject: [PATCH 14/19] build: Do proper optimization during SDCLANG_LTO

* When optimization is done at linking time,
  the optimization related flags which
  were passed to the compiler aren't respected.
 -> Introduce LOCAL_SDCLANG_LTO_LDFLAGS which
    represent the optimization related flags
    which get passed to the linker during LTO
    per module.
* If LOCAL_SDCLANG_LTO_LDFLAGS isn't set,
  we assume that the module relies on the
  optimization cflags provided by our platform
  as "default". In this case, we just inherit
  SDCLANG_COMMON_FLAGS from platform to LDFLAGS.

Change-Id: I95ecb4b9d2741a14ae39a9c674d2529c6689ada4
Signed-off-by: Alex Naidis <alex.naidis@linux.com>
Signed-off-by: Joe Maples <joe@frap129.org>
---
 core/binary.mk     | 7 ++++++-
 core/clear_vars.mk | 2 ++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/core/binary.mk b/core/binary.mk
index 63413484c..d495c4aee 100644
--- a/core/binary.mk
+++ b/core/binary.mk
@@ -550,8 +550,13 @@ my_target_global_ldflags := $($(LOCAL_2ND_ARCH_VAR_PREFIX)CLANG_$(my_prefix)GLOB
 
         ifeq ($(LOCAL_SDCLANG_LTO), true)
         ifneq ($(LOCAL_MODULE_CLASS), STATIC_LIBRARIES)
+
+            ifeq ($(strip $(LOCAL_SDCLANG_LTO_LDFLAGS)),)
+                LOCAL_SDCLANG_LTO_LDFLAGS := $(SDCLANG_COMMON_FLAGS)
+            endif
+
             SDCLANG_PRECONFIGURED_FLAGS += -fuse-ld=qcld -flto
-            my_target_global_ldflags += -fuse-ld=qcld -flto
+            my_target_global_ldflags += -fuse-ld=qcld -flto $(LOCAL_SDCLANG_LTO_LDFLAGS)
         endif
         endif
         my_target_global_cflags += $(SDCLANG_COMMON_FLAGS) $(SDCLANG_PRECONFIGURED_FLAGS)
diff --git a/core/clear_vars.mk b/core/clear_vars.mk
index 71233555c..d87f17fe0 100644
--- a/core/clear_vars.mk
+++ b/core/clear_vars.mk
@@ -446,6 +446,8 @@ LOCAL_MODULE_SYMLINKS_64:=
 LOCAL_SDCLANG:=
 LOCAL_SDCLANG_2:=
 LOCAL_SDCLANG_LTO:=
+LOCAL_SDCLANG_LTO_LDFLAGS:=
+LOCAL_SDCLANG_LTO_UNSAFE_FILTER:=
 LOCAL_SHARED_LIBRARIES_32:=
 LOCAL_SHARED_LIBRARIES_64:=
 LOCAL_SRC_FILES_32:=
-- 
2.19.1

